stl/inc/ranges

```diff
@@ -2739,6 +2968,10 @@ namespace ranges {
                     return remove_cvref_t<_Rng>{};
                 } else if constexpr (_Strat == _St::_Take_view) {
                     return take_view(_STD forward<_Rng>(_Range), _Count);
+#if _HAS_CXX23
+                } else if constexpr (!sized_range<_Rng> && _Strat == _St::_Reconstruct_repeat_view) {
+                    return repeat_view(_STD forward_like<_Rng>(*_Range._Value), _Count);

```

Comment:
Usage of `forward_like` introduces non-standard behaviour:

```c++
#include <cassert>
#include <iostream>
#include <ranges>
#include <string>

using namespace std;

int main() {
    auto v = views::repeat("text"s);
    static_cast<void>(std::move(v) | views::take(1));
    assert(*v.begin() == "text"sv); // OK on GCC (https://godbolt.org/z/vGPMY7EGe), failure with this PR.
}

```

Reply:
Is there a reason it shouldn't? i.e should there be an LWG issue to change it to forward?

Reply:
I remember why I didn't follow the exact wording in the paper now. P2494R2 allowed for move only types in range adaptors. By not forwarding that would mean you couldn't `take` or `drop` move only repeat views. Unless I'm missing something?

Reply:
> By not forwarding that would mean you couldn't take or drop move only repeat views. Unless I'm missing something?

Good point, that's true.
Personally, I think this is a good idea to use `forward_like` here, but unfortunately this is non-standard.

> i.e should there be an LWG issue to change it to forward?

I think so, that's good idea.

Reply:
Why not simply use `std::forward` like `*(_STD forward<_Rng>(_Range)._Value))`?

Reply:
I just thought it was clearer and easier to read 🤷‍♂️.

Reply:
> I agree that `forward_like` is a good idea. We should comment this with the number of the pertinent LWG issue (`// per LWG-XXXX`). Has someone submitted an issue yet, or shall I? (Also on 2999.)

Is it really an LWG? The wording of the current draft is `views​::​repeat(*E.value_, static_cast<D>(F))`, doesn’t `E` here already imply `std::forward`? And for non-copyable lvalue `view`, it does not model `viewable_range` so it will not reach here.

Reply:
Ok, sounds like the current code conforms to the WP. Please reopen if I've misunderstood the status of this comment. :smile_cat:

Reply:
The gist of this chain of comments is that the behavior implemented here is correct, and the OP actually demonstrates a bug in libstdc++. Per [[range.take.overview]/2.4]https://eel.is/c++draft/range.take.overview#2.4), `views::take(std::move(v), 1)` is expression-equivalent to `views::repeat(*std::move(v).value_, 1)`. For an object `v` with member `value_`, `std::move(v).value_` is a prvalue. `*x` for an rvalue `x` of type `std::optional<T>` is an xvalue of type `T`. The `string` stored in the _`movable-box`_ in the `repeat_view` `v` should be moved into the `repeat_view` result of the view composition expression. (FWIW, I tested with a much longer string to ensure the result in the OP isn't due to move being equivalent to copy for a string that fits in the SSO.)

Reply:
> For an object `v` with member `value_`, `std::move(v).value_` is a prvalue.

Oh, I assumed it would be a lvalue 🤦‍♂️. I'm going to report libstdc++ bug then.

Reply:
FYI, I lied about prvalue: `std::move(v).value_` is an _xvalue_ when `value_` is a non-static data member. My test program (https://godbolt.org/z/GYEKzr384) was inadequate: it's asking for the declared type of the non-static data member instead of asking for the type-and-value-category of the member access expression (https://godbolt.org/z/a514f1r7r).